src/commands/coverage/upload.ts

```diff
@@ -0,0 +1,356 @@
+    this.logger.setShouldIncludeTime(this.verbose)
+
+    if (this.flush === undefined && this.basePaths === undefined) {
+      this.context.stderr.write('Either --path or --flush must be provided\n')
+
+      return 1
+    }
+
+    await this.uploadCodeCoverageReports()
+
+    if (!this.skipGitMetadataUpload) {
+      await this.uploadGitMetadata()
+    } else {
+      this.logger.debug('Not syncing git metadata (skip git upload flag detected)')
+    }
+
+    if (!this.dryRun) {
+      this.context.stdout.write(renderSuccessfulUploadCommand(this.basePaths, this.flush))
+    }
+  }
+
+  private async uploadCodeCoverageReports() {
+    if (this.basePaths) {
+      // Normalizing the basePath to resolve .. and .
+      // Always using the posix version to avoid \ on Windows.
+      this.basePaths = this.basePaths.map((basePath) => path.posix.normalize(basePath))
+    }
+
+    this.logger.info(renderCommandInfo(this.basePaths, this.flush, this.dryRun))
+
+    const api = this.getApiHelper()
+    const payloads = await this.generatePayloads()
+
+    let fileCount = 0
+    let flushed = false
+
+    const initialTime = new Date().getTime()
+    for (const payload of payloads) {
+      fileCount += payload.paths.length
+      flushed = flushed || payload.flush
+      await this.uploadCodeCoverageReport(api, payload)
+    }
+    const totalTimeSeconds = (Date.now() - initialTime) / 1000
+
+    this.logger.info(renderSuccessfulUpload(this.dryRun, fileCount, flushed, totalTimeSeconds))
+  }
+
+  private getApiHelper(): APIHelper {
+    if (!this.config.apiKey) {
+      this.logger.error(
+        `Neither ${chalk.red.bold('DATADOG_API_KEY')} nor ${chalk.red.bold('DD_API_KEY')} is in your environment.`
+      )
+      throw new Error('API key is missing')
+    }
+
+    return apiConstructor(intakeUrl, this.config.apiKey)
+  }
+
+  private async generatePayloads(): Promise<Payload[]> {
+    const spanTags = await this.getSpanTags()
+    const customTags = this.getCustomTags()
+    const customMeasures = this.getCustomMeasures()
+
+    let reports: {format: string; paths: string[]}[] = []
+    if (this.basePaths) {
+      reports = await this.getMatchingCoverageReportFilesByFormat()
+    }
+
+    const payloads: Payload[] = []
+    if (reports.length) {
+      for (let i = 0; i < reports.length; i++) {
+        const {format, paths} = reports[i]
+        for (let j = 0; j < paths.length; j += MAX_REPORTS_PER_REQUEST) {
+          const chunkedPaths = paths.slice(j, j + MAX_REPORTS_PER_REQUEST)
+          payloads.push({
+            format,
+            paths: chunkedPaths,
+            // send flush signal with the last payload, and only if the flush flag is set
+            flush: i === reports.length - 1 && j + MAX_REPORTS_PER_REQUEST >= paths.length ? !!this.flush : false,
+            spanTags,
+            customTags,
+            customMeasures,
+            hostname: os.hostname(),
+          })
+        }
+      }
```

Comment:
(nitpick)
if you follow my suggestion to change the returned format of `getMatchingCoverageReportFilesByFormat`, this could look like this:
```javascript
    const payloads = Object.entries(reports).flatMap(([format, paths]) => {
        const numChunks = Math.ceil(paths.length / MAX_REPORTS_PER_REQUEST);
        
        return Array.from({ length: numChunks }, (_, i) => ({
            format,
            paths: paths.slice(i * CHUNK_SIZE, (i + 1) * CHUNK_SIZE),
            spanTags,
            customTags,
            customMeasures,
            hostname: os.hostname(),
        }));
    });
    // to set the last payload to flush 
    payloads[payloads.length - 1].flush = !!this.flush
```

Reply:
Updated